### PR TITLE
Fix #280 Django thread-local connection cleanup in multi threads

### LINKED ISSUE #280: Django thread-local connection cleanup in multi threads
This issue illustrates a potential issues that may affect Django adapter users.

Django's ORM establishes a database connection per thread. For the threads managed by Django (which are supposed to be used for directly handing web requests), Django ORM cleans the established connections up under the hood once the processing on the thread completes (= request_finished signal).

On the other hand, in the case where Django ORM models are used in an unmanaged thread, Django does not handle the cleanup of the connections bound to the thread although the Django framework automatically associates a new connection to the thread. This half-heartedly maintained resource possibly may cause stale DB connection related issues.

To learn more details of this Django behavior, checking the following ticket is helpful for understanding this:
https://code.djangoproject.com/ticket/9878

As Bolt for Python utilizes threads for enabling developers to run `ack()` method asynchronously (as long as `process_before_response=False`, which is the default), this framework should have a special treatment for this possible issue with Django.

As the solution for this issue (and possible similar needs), we'll introduce a new completion callback to the `ListenerRunner` mechanism. The handler will be quite similar to the `ListenerErrorHandler` callback but the given callback function is executed in the `finally` clause of listener runner invocation: 
https://github.com/slackapi/bolt-python/blob/v1.4.4/slack_bolt/listener/thread_runner.py#L99-L126
Also, for this Django specific issue, we'll add a custom lazy listener runner to the Django adapter module.

 I've already implemented the initial version of this. We can discuss its details in my upcoming pull request.

### PR DESCRIPTION
**Pull request by seratch**
This pull request resolves #280 . For the context and details of the issue we're going to address, refer to the issue #280. I will leave a few comments for reviewers.

### Category (place an `x` in each of the `[ ]`)

* [x] `slack_bolt.App` and/or its core components
* [x] `slack_bolt.async_app.AsyncApp` and/or its core components
* [x] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [ ] Others

## Requirements (place an `x` in each `[ ]`)

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: examples/django/slackapp/migrations/0001_initial.py [hunk 1]
```diff
@@ -22,15 +22,15 @@ class Migration(migrations.Migration):
                         verbose_name="ID",
                     ),
                 ),
-                ("client_id", models.TextField()),
-                ("app_id", models.TextField()),
-                ("enterprise_id", models.TextField(null=True)),
+                ("client_id", models.CharField(max_length=32)),
```
Comment (seratch) [inline examples/django/slackapp/migrations/0001_initial.py:25]: For MySQL compatibility, I've changed the database schema for this example app

File: examples/django/slackapp/models.py [hunk 2]
```diff
@@ -100,9 +101,13 @@ def logger(self) -> Logger:
 
     def save(self, installation: Installation):
         i = installation.to_dict()
+        if is_naive(i["installed_at"]):
+            i["installed_at"] = make_aware(i["installed_at"])
```
Comment (seratch) [inline examples/django/slackapp/models.py:105]: This is an improvement for MySQL apps.

File: examples/django/slackapp/models.py [hunk 3]
```diff
@@ -249,12 +254,34 @@ def consume(self, state: str) -> bool:
...
 
 
-@app.event("app_mention")
-def event_test(body, say, logger):
+def event_test(body, say, context: BoltContext, logger):
     logger.info(body)
-    say("What's up?")
+    say(":wave: What's up?")
+
+    found_rows = list(
```
Comment (seratch) [inline examples/django/slackapp/models.py:261]: If there is no thread-bound database connection when first access to any Django models, Django automatically establishes a new connection and associates it with the thread. This example demonstrates the situation where a listener performs a database query using Django ORM in a different thread (=thread managed by Bolt, not Django).

File: examples/django/slackapp/models.py [hunk 4]
```diff
@@ -249,12 +254,34 @@ def consume(self, state: str) -> bool:
...
+        SlackInstallation.objects.filter(enterprise_id=context.enterprise_id)
```
Comment (seratch) [inline examples/django/slackapp/models.py:262]: If there is no thread-bound database connection when first access to any Django models, Django automatically establishes a new connection and associates it with the thread. This example demonstrates the situation where a listener performs a database query using Django ORM in a different thread (=thread managed by Bolt, not Django).

File: slack_bolt/adapter/django/handler.py [hunk 5]
```diff
@@ -43,9 +54,86 @@ def to_django_response(bolt_resp: BoltResponse) -> HttpResponse:
     return resp
 
 
+from django.db import connections
+
+
+def release_thread_local_connections(logger: Logger, execution_type: str):
+    connections.close_all()
```
Comment (seratch) [inline slack_bolt/adapter/django/handler.py:61]: This method releases only the connections associated with the current thread.

see also:
* https://github.com/django/django/blob/3.1.7/django/db/__init__.py#L16
* https://github.com/django/django/blob/3.1.7/django/db/utils.py#L136-L237

File: slack_bolt/app/app.py [hunk 6]
```diff
@@ -339,6 +346,10 @@ def listener_runner(self) -> ThreadListenerRunner:
         """The thread executor for asynchronously running listeners."""
         return self._listener_runner
 
+    @property
+    def process_before_response(self) -> bool:
```
Comment (seratch) [inline slack_bolt/app/app.py:350]: Added the field `process_before_response` as the Djando adapter needs to know the configuration.

File: slack_bolt/listener/thread_runner.py [hunk 7]
```diff
@@ -69,6 +73,11 @@ def run(  # type: ignore
                         response=response,
                     )
                     ack.response = response
+                finally:
+                    self.listener_completion_handler.handle(
```
Comment (seratch) [inline slack_bolt/listener/thread_runner.py:77]: You can clean any resources (e.g., thread-local Django DB connections) up here.

File: slack_bolt/listener/thread_runner.py [hunk 8]
```diff
@@ -122,6 +131,11 @@ def run_ack_function_asynchronously():
                                 response=response,
                             )
                             ack.response = response
+                    finally:
+                        self.listener_completion_handler.handle(
```
Comment (seratch) [inline slack_bolt/listener/thread_runner.py:135]: You can clean any resources (used in a lazy listener) up here.

**Review by seratch — COMMENTED** [review]
comments for reviewers

File: slack_bolt/listener/async_internals.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        "error": error,
+        "client": request.context.client,
+        "req": request,
+        "request": request,
+        "resp": response,
+        "response": response,
+        "context": request.context,
+        "body": request.body,
+        # payload
+        "body": request.body,
```
Comment (mwbrooks) [inline slack_bolt/listener/async_internals.py:35]: Looks like `body` was declared twice (I'm guessing the `#payload` comment is meant to include body):
```suggestion
        # payload
        "body": request.body,
```

File: slack_bolt/listener/internals.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        "error": error,
+        "client": request.context.client,
+        "req": request,
+        "request": request,
+        "resp": response,
+        "response": response,
+        "context": request.context,
+        "body": request.body,
+        # payload
+        "body": request.body,
```
Comment (mwbrooks) [inline slack_bolt/listener/internals.py:36]: Same suggestion as above. Please disregard if it's on purpose:
```suggestion
        # payload
        "body": request.body,
```

**Review by mwbrooks — APPROVED** [review]
Looks good. I just added minor suggestions below.

File: slack_bolt/listener/async_internals.py (hunk 9 above)
Comment (seratch) [reply slack_bolt/listener/async_internals.py]: @mwbrooks Thanks for catching these!

**Pull request merged**